### PR TITLE
centralize details, restore flags

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Centralizes details and restore flags for all 3
m365 services.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #2786

#### Test Plan

- [x] :muscle: Manual
- [x] :green_heart: E2E

### CODE REVIEW

**Comment by gmatev** [comment]
@ryanfkeepers does this preserve the ability to control the flag order?

File: src/cli/utils/sharepoint.go [hunk 1]
```diff
@@ -36,14 +42,65 @@ type SharePointOpts struct {
...
 
+// AddSharePointDetailsAndRestoreFlags adds flags that are common to both the
+// details and restore commands.
+func AddSharePointDetailsAndRestoreFlags(cmd *cobra.Command) {
+	fs := cmd.Flags()
+
+	fs.StringVar(
+		&Library,
+		LibraryFN, "",
+		"Select only this library. Default includes all libraries.")
```
Comment (gmatev) [inline src/cli/utils/sharepoint.go:53]:
```suggestion
		"Select only this library; defaults to all libraries")
```
Consistency with flag below

**Comment by ryanfkeepers** [comment]
> does this preserve the ability to control the flag order?

I have no reason to believe it wouldn't.  I can run some checks for proof afterward.

**Comment by gmatev** [comment]
the text tweaks look good to me as long as we have preserved the ordering of the flags. 
I'll let @ashmrtn commend on the other changes.

File: src/cli/backup/exchange.go [hunk 2]
```diff
@@ -108,25 +87,20 @@ func addExchangeCommands(cmd *cobra.Command) *cobra.Command {
...
 			"Select one or more types of data to backup: "+dataEmail+", "+dataContacts+", or "+dataEvents)
 		options.AddOperationFlags(c)
 
 	case listCommand:
-		c, fs = utils.AddCommand(cmd, exchangeListCmd())
-
-		fs.StringVar(&backupID,
-			"backup", "",
-			"ID of the backup to retrieve.")
+		c, _ = utils.AddCommand(cmd, exchangeListCmd())
```
Comment (ashmrtn) [inline src/cli/backup/exchange.go:99]: seems like you could remove the flag set return value from `AddCommand`?

**Review by ashmrtn — APPROVED** [review]

**Comment by ashmrtn** [comment]
maybe eventually we should not use global flags but use flags tied to an instance of a struct? This would allow us some parallelism in tests if we wanted it, though it would take a bit more work to setup

File: src/cli/backup/exchange.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/cli/backup/exchange.go:99]: it's used in other places, primarily to maintain `fs.SortFlags = false`.  We should keep it in the return set for now.  Can always revisit later.

**Comment by ryanfkeepers** [comment]
> maybe eventually we should not use global flags but use flags tied to an instance of a struct? This would allow us some parallelism in tests if we wanted it, though it would take a bit more work to setup

Yep!  Just hasn't been a hard need as of yet.

**Pull request merged**